### ISSUE TITLE
Non http Handler version of SPNEGOKRB5Authenticate

### ISSUE THREAD

**Issue by bwmarrin** [jcmturner/gokrb5#314]
Is there a way to perform the functions that SPNEGOKRB5Authenticate as a normal function rather than a http handler wrapper? 

That would return a struct of the authentication information, or an not authorized error?

It would be pretty handy and add more flexibility to how this library is used and mixed in with web services wrote in Go.

If there's not a way to do this now, I'd be happy to PR in a way of doing it.

**Comment by bwmarrin** [comment]
The above question could be a bit of an xyproblem..

What I'd like to be able to accomplish immediately, is have more control over how unauthenticated sessions are handled.  If kerberos fails, I'd like to implement fallback options.  So far I've been testing with the httpServer.go example and it seems to just send "Unauthorized" to the client and not pass the request down the stack.

So if either the above option existed where I had more control, or this handler would pass the call down the stack where you could check if it was authorized or not and then act accordingly that would work too.

**Comment by bwmarrin** [comment]
So, I've been toying around with a solution for this, and I think I've found one.

First step, is to create a new function similar to `SPNEGOKRB5Authenticate`, except instead of it being a middleware wrapper design, make it just a simple function that returns the auth status, kerberos credentials data, and an error.   I also like this because now the kerberos ticket data could be returned as a normal value instead of being stuck inside a context value and needing casting to use it.

Next, the web app can use this new function to write it's own own middleware func.  When you call the function and it returns an error or otherwise is unable to authenticate the user - return a 401 header WITH an HTML login screen. 

Doing this, the user will be automaticlly authenticated via spengo if it works - otherwise they are presented with a nice HTML login page as a fallback option.   You could also make the fallback a pretty "Unauthorized" page too if one wanted.  Or a number of other things - the key here is that the library (gokrb5) user has a lot more control on what happens when using gokrb5 in this fashion.

I've wrote a small proof of concept doing this and I'll try to clean it up and post the example.  I'm not sure what the best way to integrate this into gokrb5 is - so I'd like your recommendations @jcmturner .


This may also be a solution for https://github.com/jcmturner/gokrb5/issues/313

**Comment by bwmarrin** [comment]
Okay, I've created a function and example for this.  Let me know your thoughts.

First - add a this (or similar) function to the library 
https://github.com/bwmarrin/gokrb5/blob/master/spnego/http.go#L281

Then it can be used as below,
```go
func testAuthHandler(w http.ResponseWriter, r *http.Request) {
        kt, err := keytab.Load("my.keytab")
        if err != nil {
                lit.Error("error loading keytab file, %s", err)
                return
        }

        a, c, err := spnego.Authenticate(kt, w, r)
        if err != nil {
                fmt.Println(err)
        }

        if !a {
                http.ServeFile(w, r, "login.html")
                return
        }

        fmt.Fprint(w, "<html>\n<p><h1>TEST.GOKRB5 Handler</h1></p>\n")
        fmt.Fprintf(w, "<ul>\n")
        fmt.Fprintf(w, "<li>Authenticed user: %s</li>\n", c.UserName())
        fmt.Fprintf(w, "<li>User's realm: %s</li>\n", c.Domain())
        fmt.Fprintf(w, "<li>Authn time: %v</li>\n", c.AuthTime())
        fmt.Fprintf(w, "<li>Session ID: %v</li>\n", c.SessionID())
        fmt.Fprintf(w, "<li>displayname : %v</li>\n", c.DisplayName())
        fmt.Fprintf(w, "<li>expired: %v</li>\n", c.Expired())
        fmt.Fprintf(w, "<li>human: %v</li>\n", c.Human())
        fmt.Fprintf(w, "<li>: %v</li>\n", c.AddAuthzAttribute)
        fmt.Fprint(w, "<li>Authz Attributes (Group Memberships):</li><ul>\n")
        for _, s := range c.AuthzAttributes() {
                fmt.Fprintf(w, "<li>%v</li>\n", s)
        }
        fmt.Fprint(w, "</ul>\n")

        if ADCreds, ok := c.Attributes()[credentials.AttributeKeyADCredentials].(credentials.ADCredentials); ok {
                // Now access the fields of the ADCredentials struct. For example:
                fmt.Fprintf(w, "<li>EffectiveName: %v</li>\n", ADCreds.EffectiveName)
                fmt.Fprintf(w, "<li>FullName: %v</li>\n", ADCreds.FullName)
                fmt.Fprintf(w, "<li>UserID: %v</li>\n", ADCreds.UserID)
                fmt.Fprintf(w, "<li>PrimaryGroupID: %v</li>\n", ADCreds.PrimaryGroupID)
                fmt.Fprintf(w, "<li>Group SIDs: %v</li>\n", ADCreds.GroupMembershipSIDs)
                fmt.Fprintf(w, "<li>LogOnTime: %v</li>\n", ADCreds.LogOnTime)
                fmt.Fprintf(w, "<li>LogOffTime: %v</li>\n", ADCreds.LogOffTime)
                fmt.Fprintf(w, "<li>PasswordLastSet: %v</li>\n", ADCreds.PasswordLastSet)
                fmt.Fprintf(w, "<li>LogonServer: %v</li>\n", ADCreds.LogonServer)
                fmt.Fprintf(w, "<li>LogonDomainName: %v</li>\n", ADCreds.LogonDomainName)
                fmt.Fprintf(w, "<li>LogonDomainID: %v</li>\n", ADCreds.LogonDomainID)
        }
        fmt.Fprint(w, "</ul>")
}
```

One area here that could probably still be improved on, is - in this example the keytab is parsed on every request.  Since this is probably a hot-path for an application it might make sense to have a spnego.New() func that can parse the keytab, apply any configuration settings, etc so those only need to be done once.

**Comment by jcmturner** [comment]
I see you are looking to allow the browser to use basic authentication to send the username password (as in https://tools.ietf.org/html/rfc2617#section-2) or resort to the application to presenting a page for form based authentication.
The challenge is with the initial negotiation. The browser will most often not initially provide an SPNEGO value in the headers until it has received the first 401 informing it that negotiation is supported. Therefore I think your example may always fall back to the login page.

**Comment by bwmarrin** [comment]
Yeah, you're right. It does technically always show the login page.  

But if Kerberos works, you'll actually never see the login page because how fast the browser reacts to the 401.  You just end up being logged in.

I've been using a method very similar to my example above now for, well since November on a few apps.  It's working great.   I setup the fall back to use LDAP auth against the domain controller.

**Comment by bwmarrin** [comment]
Another solution here, would be to modify (expose) enough of the library so that the SPNEGOKRB5Authenticate method can mostly be remade outside of the library.  Then end-users can implement this on their own.  I looked at what that would take, and it's a handful of items that would have to be exposed.  So I'm not sure if that makes things to messy - and a end using re-implementing that func might be a bit too much for most.

The biggest change that has to be made to the current one, for this to work, is the removal of this line, https://github.com/jcmturner/gokrb5/blob/913509a6da7657b65e1d039f7089a52299f1079a/spnego/http.go#L225

Having that there it prevents the http request from continuing so a fallback option can be presented. 

Not sure what the best fix is and I'm open to helping test or whatever if I can.  I'm using my fork of the library right now so I can do this fallback auth - and I'd like to switch back to using the official library instead :)

**Comment by m-vinc** [comment]
I know that issue is old but I actually want to customize the return message when a negotiation failed, like returning json instead of the simple Unauthorized message the library actually write using the writer of the request.
I created a fork to pass a custom handler called when an unauthorized is returned but your approach is much better.

I've reuse your `Authenticate` function I just  change the context key from`CTXKeyCredentials`to`ctxCredentials` and remove the `.get` of the `CTXKeyAuthenticated` context value since this key doesn't exist anymore I guess ?

**Comment by bwmarrin** [comment]
Oh, hey there :)  Yeah this never got much traction here :(.  I am still using my above code in production on multiple systems with hundreds of clients using it daily flawlessly.  I however have not upgraded my fork of this library in quite awhile so I'm not sure what changes need to be made for it to work with the latest code here.  I definitively should update my fork and custom function.  I was really hoping something would get changed in this library so I wouldn't need to keep my own fork but the functionality is 100% essential for my apps.  Is an accommodation for this feature ever going to be considered @jcmturner?  Or should the issue be closed? No sense in it sitting open for years if there's no chance of implementing something for it.

**Comment by bwmarrin** [comment]
@m-vinc to further answer you specific questions - let me look at the current code here and see what needs done and I'll try to write back this week.
